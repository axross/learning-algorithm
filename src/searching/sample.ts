export const sortedCharactors = [
  "a",
  "b",
  "c",
  "d",
  "e",
  "f",
  "g",
  "h",
  "i",
  "j",
  "k",
  "l",
  "m",
  "n",
  "o",
  "p",
  "q",
  "r",
  "s",
  "t",
  "u",
  "v",
  "w",
  "x",
  "y",
  "z"
];

export const unsortedCharactors = [
  "c",
  "n",
  "r",
  "j",
  "m",
  "s",
  "b",
  "u",
  "w",
  "y",
  "h",
  "t",
  "v",
  "f",
  "z",
  "g",
  "x",
  "o",
  "k",
  "q",
  "d",
  "e",
  "p",
  "l",
  "a",
  "i"
];
