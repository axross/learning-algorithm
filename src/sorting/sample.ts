import roundRobin from "../test-utility/round-robin.ts";
import randomInteger from "../test-utility/random-integer.ts";

export const charactors = ["a", "b", "c", "d", "e", "f", "g", "h"];

export const samples = roundRobin(charactors).slice(1);

export function getRandomSample() {
  return samples[randomInteger(0, samples.length - 1)];
}

export const staticSample = ["f", "a", "h", "b", "d", "g", "e", "c"];

export const charactorPairs = ["a0", "a1", "b0", "b1", "c0", "c1", "d0", "d1"];

export const charactorPairSamples = roundRobin(["a0", "b0", "c0", "d0"]).reduce(
  (arr, pattern0) => [
    ...arr,
    ...roundRobin(["a1", "b1", "c1", "d1"]).reduce(
      (arr, pattern1) => [...arr, [...pattern0, ...pattern1]],
      [] as string[][]
    )
  ],
  [] as string[][]
);

export const randomCharactorPairSample =
  charactorPairSamples[randomInteger(0, charactorPairSamples.length - 1)];

export const staticCharactorPairSample =
  randomCharactorPairSample[Math.floor((samples.length / 7) * 5)];
