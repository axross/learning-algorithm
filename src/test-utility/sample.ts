function shuffle<Value>(arr: Value[]): Value[] {
  const shuffled = [...arr];

  for (let i = shuffled.length - 1; i >= 0; i--) {
    const rand = Math.floor(Math.random() * (i + 1));

    [shuffled[i], shuffled[rand]] = [shuffled[rand], shuffled[i]];
  }

  return shuffled;
}

export const sampleCharactors = [
  ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 65)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97))
];

export const shuffledSampleCharactors = shuffle(sampleCharactors);

export const sampleTwoCharactors = [
  ...Array.from({ length: 5 }, (_, i) => [
    String.fromCharCode(i + 48) + "<",
    String.fromCharCode(i + 48) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 65) + "<",
    String.fromCharCode(i + 65) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 97) + "<",
    String.fromCharCode(i + 97) + ">"
  ])
].reduce((array, value) => [...array, ...value], []);

export const shuffledSampleTwoCharactors = [
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + "<")
  ]),
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + ">")
  ])
];
