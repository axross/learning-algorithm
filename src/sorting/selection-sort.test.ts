import selectionSort, {
  SelectionSortComparison,
  SelectionSortSwap
} from "./selection-sort";
import { charactors, randomSample, staticSample } from "./sample";

test(`selectionSort(collection: ["${randomSample[0]}", "${randomSample[1]}", "${
  randomSample[2]
}" ... ${
  randomSample.length
} items]) sorts the collection to be ordered`, () => {
  const collection = [...randomSample];

  selectionSort({
    collection,
    compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0)
  });

  expect(collection).toEqual(charactors);
});

test("the steps of selectionSort() matches the previous snapshot", () => {
  const comparisons: SelectionSortComparison<string>[] = [];
  const swaps: SelectionSortSwap<string>[] = [];
  const collection = [...staticSample];

  selectionSort({
    collection,
    compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0),
    onCompare: comparison => comparisons.push(comparison),
    onSwap: swap => swaps.push(swap)
  });

  expect({
    from: staticSample,
    to: collection,
    comparisons,
    swaps
  }).toMatchSnapshot();
});
