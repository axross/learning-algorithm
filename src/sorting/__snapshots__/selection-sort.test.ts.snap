// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the steps of selectionSort() matches the previous snapshot 1`] = `
Object {
  "comparisons": Array [
    Object {
      "a": Object {
        "index": 0,
        "value": "g",
      },
      "b": Object {
        "index": 0,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 0,
        "value": "a",
      },
      "b": Object {
        "index": 0,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 0,
        "value": "b",
      },
      "b": Object {
        "index": 2,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 0,
        "value": "c",
      },
      "b": Object {
        "index": 2,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 0,
        "value": "d",
      },
      "b": Object {
        "index": 2,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 0,
        "value": "e",
      },
      "b": Object {
        "index": 2,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 0,
        "value": "h",
      },
      "b": Object {
        "index": 2,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 1,
        "value": "f",
      },
      "b": Object {
        "index": 1,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 1,
        "value": "b",
      },
      "b": Object {
        "index": 2,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 1,
        "value": "c",
      },
      "b": Object {
        "index": 3,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 1,
        "value": "d",
      },
      "b": Object {
        "index": 3,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 1,
        "value": "e",
      },
      "b": Object {
        "index": 3,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 1,
        "value": "h",
      },
      "b": Object {
        "index": 3,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "g",
      },
      "b": Object {
        "index": 2,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "c",
      },
      "b": Object {
        "index": 2,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "d",
      },
      "b": Object {
        "index": 4,
        "value": "c",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "e",
      },
      "b": Object {
        "index": 4,
        "value": "c",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "h",
      },
      "b": Object {
        "index": 4,
        "value": "c",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "f",
      },
      "b": Object {
        "index": 3,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "d",
      },
      "b": Object {
        "index": 4,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "e",
      },
      "b": Object {
        "index": 5,
        "value": "d",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "h",
      },
      "b": Object {
        "index": 5,
        "value": "d",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "g",
      },
      "b": Object {
        "index": 4,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "e",
      },
      "b": Object {
        "index": 4,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "h",
      },
      "b": Object {
        "index": 6,
        "value": "e",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "f",
      },
      "b": Object {
        "index": 5,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "h",
      },
      "b": Object {
        "index": 6,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 6,
        "value": "h",
      },
      "b": Object {
        "index": 6,
        "value": "g",
      },
    },
  ],
  "from": Array [
    "f",
    "g",
    "a",
    "b",
    "c",
    "d",
    "e",
    "h",
  ],
  "swaps": Array [
    Object {
      "a": Object {
        "index": 0,
        "value": "f",
      },
      "b": Object {
        "index": 2,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 1,
        "value": "g",
      },
      "b": Object {
        "index": 3,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "f",
      },
      "b": Object {
        "index": 4,
        "value": "c",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "g",
      },
      "b": Object {
        "index": 5,
        "value": "d",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "f",
      },
      "b": Object {
        "index": 6,
        "value": "e",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "g",
      },
      "b": Object {
        "index": 6,
        "value": "f",
      },
    },
  ],
  "to": Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
    "h",
  ],
}
`;
